finetrainers/trainer.py

```diff
@@ -0,0 +1,542 @@
+        # Afterwards we recalculate our number of training epochs
+        self.state.train_epochs = math.ceil(self.state.train_steps / num_update_steps_per_epoch)
+    
+    def prepare_trackers(self) -> None:
+        logger.info("Initializing trackers")
+
+        tracker_name = self.args.tracker_name or "finetrainers-experiment"
+        self.state.accelerator.init_trackers(tracker_name, config=self.args.to_dict())
+    
+    def train(self) -> None:
+        logger.info("Starting training")
+
+        memory_statistics = self.get_memory_statistics()
+        logger.info(f"Memory before training start: {json.dumps(memory_statistics, indent=4)}")
+
+        self.state.train_batch_size = self.args.batch_size * self.state.accelerator.num_processes * self.args.gradient_accumulation_steps
+        info = {
+            "trainable parameters": self.state.num_trainable_parameters,
+            "total samples": len(self.dataset),
+            "train epochs": self.state.train_epochs,
+            "train steps": self.state.train_steps,
+            "batches per device": self.args.batch_size,
+            "total batches observed per epoch": len(self.dataloader),
+            "train batch size": self.state.train_batch_size,
+            "gradient accumulation steps": self.args.gradient_accumulation_steps,
+        }
+        logger.info(f"Training configuration: {json.dumps(info, indent=4)}")
+
+        # TODO(aryan): handle resume from checkpoint
+
+        global_step = 0
+        first_epoch = 0
+        initial_global_step = 0
+        progress_bar = tqdm(range(0, self.state.train_steps), initial=initial_global_step, desc="Training steps", disable=not self.state.accelerator.is_local_main_process)
+
+        accelerator = self.state.accelerator
+        weight_dtype = self.state.weight_dtype
+        transformer_config = self.transformer.config
+        scheduler_sigmas = self.scheduler.sigmas.clone().to(device=accelerator.device, dtype=weight_dtype)
+        generator = torch.Generator(device=accelerator.device)
+        if self.args.seed is not None:
+            generator = generator.manual_seed(self.args.seed)
+        
+        for epoch in range(first_epoch, self.state.train_epochs):
+            logger.debug(f"Starting epoch ({epoch + 1}/{self.state.train_epochs})")
+            
+            self.transformer.train()
+            models_to_accumulate = [self.transformer]
+
+            for step, batch in enumerate(self.dataloader):
+                logger.debug(f"Starting step ({step + 1})")
+                logs = {}
+
+                with accelerator.accumulate(models_to_accumulate):
+                    videos = batch["videos"]
+                    prompts = batch["prompts"]
+                    batch_size = len(prompts)
+
+                    if self.args.caption_dropout_technique == "empty":
+                        if random.random() < self.args.caption_dropout_p:
+                            prompts = [""] * batch_size
```

Comment:
Note that different models have different ways to incorporate this. For example, in Mochi-1, we directly [zero out the embeddings](https://github.com/genmoai/mochi/blob/main/demos/fine_tuner/train.py#L292C13-L293C39). 

So, `encode_prompt(prompt="")` is NOT the same as doing `out = encode_prompt(prompt="my dog"); out.zero_()`.

Reply:
awesome, thanks